.github/copilot-instructions.md

```diff
@@ -0,0 +1,314 @@
+    cmd.on_failure { |errors| render_400(:failed_validations, errors: errors) }
+  end
+end
+```
+
+Key characteristics:
+- Requires authentication via Bearer tokens (`Authorization: Bearer <token>`)
+- Returns JSON responses
+- Handles errors consistently via `render_400`, `render_403`, etc.
+- Routes defined in `config/routes/api.rb`
+
+#### SPI Routes (`/spi`)
+The Service Provider Interface (SPI) provides internal endpoints for:
+- AWS Lambda functions
+- Internal microservices
+- Exercism infrastructure components
+
+**Security Model**: SPI endpoints are secured at the AWS infrastructure level rather than application-level authentication, allowing trusted internal services to post data arbitrarily.
+
+```ruby
+# Example SPI endpoints
+namespace :spi do
+  resources :tooling_jobs, only: :update        # Tooling service updates
+  resources :chatgpt_responses, only: :create   # AI service responses
+  get "solution_image_data/:track_slug/:exercise_slug/:user_handle" => "solution_image_data#show"
+end
+```
+
+Key characteristics:
+- No application-level authentication required
+- AWS-level security controls access
+- Used by Lambda functions to post results back to main application
+- Routes defined in `config/routes/spi.rb`
+
+#### Standard Routes
+Regular Rails routes handle:
+- User-facing web pages
+- Authentication flows (Devise)
+- Webhooks (GitHub, Stripe, PayPal)
+- Admin interfaces
+
+Routes are organized in the main `config/routes.rb` with additional route files for specific features like bootcamp functionality.
+
+### Configuration
+- `config/database.yml` - Database configuration
+- `Procfile.dev` - Development server orchestration
+- `.dockerimages.json` - Docker service versions
+- `package.json` - JavaScript dependencies and scripts
+- `Gemfile` - Ruby dependencies
+
+### Build System
+- `app/javascript/esbuild.js` - JavaScript build configuration
+- `postcss.config.js` - CSS build configuration
+- `tailwind.config.js` - Tailwind CSS customization
+
+## Common Tasks and Debugging
+
+### Database Issues
+- Reset database: `bundle exec rails db:drop db:create db:migrate db:seed`
+- Check migrations: `bundle exec rails db:migrate:status`
+
+### Asset Issues  
+- Clear assets: `rm -rf .built-assets/`
+- Rebuild: `yarn build:css && yarn build`
+
+### Service Issues
+- Check Docker services: `docker ps`
+- Restart LocalStack: `docker restart <localstack-container-id>`
+- Check AnyCable: `bundle exec anycable --version`
+
+### Common Error Patterns
+- "Ruby version mismatch": Install Ruby 3.4.4 exactly
+- "NPM authentication failed": Need NPM_TOKEN for private packages
+- "Database connection failed": Check MySQL service and credentials
+- "Asset compilation failed": Ensure Node.js and Yarn versions are compatible
+
+This setup is complex but necessary for the full Exercism platform. When in doubt, refer to the CI configuration in `.github/workflows/tests.yml` for authoritative build commands and timing expectations.
```

Comment:
Add a section somewhere about testing. Determine normal testing patterns for models, commands and system tests and write a section for each. We normally try and have lots of small tests for every public model method. For commands, we try and have a happy path test, and then every possible edge-case test. Normally in the form of setup, TheCommand.(...), an assertion.